### ISSUE TITLE
RFC Tracker: Static Evaluation of Provider Iteration

### ISSUE THREAD

**Issue by apparentlymart** [opentofu/opentofu#2123]
This is the primary tracking issue for the work described in **[Static Evaluation of Provider Iteration](https://github.com/opentofu/opentofu/blob/main/rfc/20240513-static-evaluation-providers.md)**, which was written as part of https://github.com/opentofu/opentofu/pull/1649.

Related feature requests:
- #300

The work represented by this issue is intended to be completed for the OpenTofu v1.9.0 release.

---

> [!NOTE]  
> **What is an "RFC tracking issue" anyway?**
>
> This issue is intended as a trial-run of the process described in a not-yet-approved process RFC proposed in https://github.com/opentofu/opentofu/pull/2100.
>
> Because that process change has not yet been adopted, and because this tracking issue was created retroactively for an RFC whose work was already in progress, this particular tracking issue is likely to be used a little inconsistently and might not follow the described process exactly; @apparentlymart is maintaining it on a best-effort basis to get some experience with how tracking issues might work in practice and learn from any questions or challenges that arise.
>
> Hopefully this does also serve a useful direct purpose as a tracking issue, but I apologize in advance if anything goes amiss here that makes this a less-than-accurate reflection of the project.

# Feature-completion Tasks

The following are broad items of work we'd need to finish before work on this RFC is "feature complete" and thus ready for enthusiastic testing, though not necessarily yet release-ready (there will probably be some bugs).

This set of items is intended to align with the first "beta" release. We will probably make an alpha release before this list is 100% resolved to request early feedback on the overall design and primary behavior. This list can keep growing until we reach the first beta, after which we'll switch to using the "Final release-blocking Tasks" section below.

- [x] (Various earlier work on this project is not captured here because it significantly predates the creation of this tracking issue. Sorry! At least some of it is represented by https://github.com/opentofu/opentofu/pull/2054 .)
- [x] Explore what constraints we ought to keep in mind to leave room for fully-dynamic provider configuration expansion later: https://github.com/opentofu/opentofu/pull/2088

    (Merging the other RFC PR is not a requirement for this item to be completed, only that we've developed it enough to decide what changes we want to make to _this_ project's RFC for the next item.)
- [x] Revise the initial RFC to include what we learned from the dynamic-expansion exploration: https://github.com/opentofu/opentofu/pull/2115

    @apparentlymart is driving this.
- [x] Implementation of the main functionality described in the revised PR:
    - https://github.com/opentofu/opentofu/pull/2105
    - https://github.com/opentofu/opentofu/pull/2111
    - https://github.com/opentofu/opentofu/pull/2121

    @cam72cam is driving these final implementation tasks.
- [x] v1.9.0-alpha1 and v1.9.0-alpha2 feedback
    - [x] https://github.com/opentofu/opentofu/pull/2152 (for https://github.com/opentofu/opentofu/issues/300#issuecomment-2462533054)
    - [x] https://github.com/opentofu/opentofu/issues/2149
- [x] https://github.com/opentofu/opentofu/pull/2186
- [x] Early/internal testing: OpenTofu Core team members and interested+adventurous community members test using either builds from `main` or an alpha release to give early feedback. (Could potentially add new items to this list, or to the other lists below.)
    - [x] Invite those watching #300 to participate.
        https://github.com/opentofu/opentofu/issues/300#issuecomment-2462708057
    - [x] Post the feature preview video. (alpha) @abstractionfactory 
    - [x] ["Help us test" blog post](https://opentofu.org/blog/help-us-test-opentofu-1-9-0-alpha2/)
- [x] First round of reference documentation updates; at least enough for those participating in beta feedback to understand how this works without having to dig into the RFC.
    - https://github.com/opentofu/opentofu/pull/2138

    @apparentlymart is driving the documentation updates.

# Final release-blocking Tasks

The following are items that must be completed before the final release of these features. This list can grow during the prerelease testing period.

- [x] https://github.com/opentofu/opentofu/issues/2269
- [x] https://github.com/opentofu/opentofu/issues/2277
- [x] https://github.com/opentofu/opentofu/issues/2268
- [x] https://github.com/opentofu/opentofu/pull/2284
- [x] Changelog entry under "New Features"
- [x] Finalized reference documentation
    - We didn't get any significant feedback on the first round of docs, so we'll just ship with that and iterate later if necessary.
- [x] https://github.com/opentofu/opentofu/issues/2334
    - [x] https://github.com/opentofu/opentofu/pull/2335
- [x] https://github.com/opentofu/opentofu/issues/2342
- [x] Release announcement content:
    - Release blog post @abstractionfactory 

# Post-release tasks

The following tasks could potentially wait until after final release, but should still get done. (There's no harm in doing items from this list _before_ the final release instead, where possible.)

- [x] Post the release blog post.
    - Done: https://opentofu.org/blog/opentofu-1-9-0/
- [x] Create pre-emptive feature request issues for each of the likely future requests ("missing features") we identified in the RFC, to encourage emoji-voting and other feedback on them to be created in a central place.

# Deferred Tasks

The following are tasks that were previously in one of the lists above but we have intentionally taken them out of scope for this first round of work. If this issue is closed before these are all resolved then we should ensure that each unresolved one has an open issue to track it as future work.

- Support for `for_each` in `provider` blocks in text editor and IDE integrations.

    Currently some populate integrations mark `for_each` in a `provider` block as an error, which is likely to discourage its use. Since the current integrations were created for this project's predecessor rather than for OpenTofu, a current workaround is to put the `provider` block in a `.tofu` file instead of a `.tf` file so that an OpenTofu-unaware integration won't check it for errors _at all_, which is far from ideal but perhaps less distracting than a spurious error. (https://github.com/opentofu/opentofu-ls/issues/9)
- [Make the provider instance key handling convert to `cty.String` rather than _require_ `cty.String`](https://github.com/opentofu/opentofu/pull/2105#discussion_r1817183248), for consistency with how instance key lookups work elsewhere in the language. (https://github.com/opentofu/opentofu/issues/2378)
- Fix some little naming inconsistencies in the state snapshot read/write code, as discussed in [a comment on the main PR](https://github.com/opentofu/opentofu/pull/2105#discussion_r1826371047). (https://github.com/opentofu/opentofu/issues/2379)
- https://github.com/opentofu/opentofu/issues/2155
- Address some of the naming confusion that came from earlier work, as discussed in [a comment on the main PR](https://github.com/opentofu/opentofu/pull/2105#discussion_r1827713572). (https://github.com/opentofu/opentofu/issues/2379)

**Comment by apparentlymart** [comment]
Our revised RFC includes some slightly-finicky rules for how exactly we record the relationship between resource instance and provider instance in state snapshots, so that the new version of OpenTofu can still produce state snapshots that are backward-compatible with older versions as long as the configuration isn't using this new feature.

However, I think there's an awkward edge-case that we should be sure to test and make sure that it behaves in a sensible way:

1. Start by planning and applying the following configuration to create an initial state snapshot:

    ```hcl
    locals {
      instances = toset(["a", "b"])
    }

    provider "null" {
      alias    = "test"
      for_each = local.instances
    }

    resource "null_resource" "test" {
      for_each = local.instances
      provider = null.test[each.key]
    }
    ```

    This should cause the final state snapshot to track that `null_resource.test["a"]` belongs to `null.test["a"]` and that `null_resource.test["b"]` belongs to `null.test["b"]`.

2. Now modify the configuration as follows, changing the provider configuration's alias but otherwise keeping everything else the same:

    ```hcl
    locals {
      instances = toset(["a", "b"])
    }

    provider "null" {
      alias    = "test2"
      for_each = local.instances
    }

    resource "null_resource" "test" {
      for_each = local.instances
      provider = null.test2[each.key]
    }
    ```

    *But* when planning and applying this, use `-target='null_resource.test["a"]'` so that the graph walk won't visit `null_resource.test["b"]`.

    What ought to happen here is perhaps debatable, but the worst-case scenario we'd want to _avoid_ is having the state snapshot entry for `null_resource.test["b"]` still be associated with `aws.test["b"]` even though `null_resource.test["a"]` is now associated with `aws.test2["a"]`, because that would violate our consistency rule that all instances of a given resource must belong to the same provider configuration.

3. Just to force the consequence of whatever happened in the previous step, now delete the `resource "null_resource" "test"` block from the configuration altogether and try to create a plan.

    At this point we no longer have anything in the configuration to use to decide which provider instance is responsible for planning and applying the deletion of the two instances of `null_resource.test`, so OpenTofu will need to rely on the state to decide that.

    One way we could fail at this point is that the previous step might've generated a state snapshot that the state snapshot loader refuses to load because it's inconsistent. I think that outcome would be _very bad_ because it would effectively prevent any future work on the same workspace until someone performs state surgery to make the state snapshot valid again.

    Another possible failure mode that would be more subtle is for the plan graph to have incorrect dependency edges because the plan graph builder doesn't know how to make sense of a resource with instances belonging to different provider configurations. This one might be trickier to notice because it could result in successfully creating an incomplete dependency graph, and therefore the failure might not be reliably reproducible.

Once we've nailed down exactly how the above ought to behave I suggest that we add a context test that covers it, since this seems like something that would be relatively easy to accidentally regress in future given how `-target` is handled very far away from the recording of the bindings between resource instances and provider instances.

**Comment by apparentlymart** [comment]
This work has now all been released, and follow-up issues created from the list of deferred tasks, so we're all done here. 🎉
